driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java

```diff
@@ -60,11 +63,17 @@ public void startTransaction() {
 
     @Override
     public void startTransaction(final TransactionOptions transactionOptions) {
+        notNull("transactionOptions", transactionOptions);
         if (inTransaction) {
             throw new IllegalStateException("Transaction already in progress");
         }
         inTransaction = true;
-        this.transactionOptions = transactionOptions;
+        TransactionOptions.Builder newOptionsBuilder = TransactionOptions.builder();
```

Comment:
One particularly ugly aspect of this is that the options we apply to the session are not actually the options that were supplied.  Instead they are the result of merging the options with the default options, which have in turn been merged with the MongoClient options.

Reply:
I would encapsulate all this logic into the builder:

`builder(transactionOptions, defaultTransactionOptions);`

Or even a factory method:

`mergeTransactionOptions(transactionOptions, defaultTransactionOptions);`

Makes it easier to test and extend in the future. Also could DRY up the code a little, the merge logic seems to be duplicated in the `createClientSession` method.

Reply:
Added TransactionOptions.merge factory method